kge/model/transr.py

```diff
@@ -0,0 +1,117 @@
+import torch
+from kge import Config, Dataset
+from kge.model.kge_model import RelationalScorer, KgeModel
+from torch.nn import functional as f
+
+
+class TransRScorer(RelationalScorer):
+    r"""Implementation of the TransR KGE scorer."""
+
+    def __init__(self, config: Config, dataset: Dataset, configuration_key=None):
+        super().__init__(config, dataset, configuration_key)
+        self._norm = self.get_option("l_norm")
+
+    def _transfer(self, ent_emb, rel_emb, projection_matrix):
+        n = rel_emb.size(0)
+        k = ent_emb.size(1)
+        d = rel_emb.size(1)
+
+        ent_emb = ent_emb.view(n, k, 1)
+        projection_matrix = projection_matrix.view(n, d, k)
+
+        out = torch.matmul(projection_matrix, ent_emb)
+
+        return out.reshape(-1, d)
+
+    def score_emb(self, s_emb, p_emb, o_emb, combine: str):
+        test = torch.chunk(p_emb, 129, dim=1)
+
+        rel_emb = test[0]
+        projection_matrix = torch.cat(test[-(len(test)-1):])
```

Comment:
The way the projection matrix is obtained feels unclean at best, what's your opinion about that? (Also see line 116)

Reply:
I don't think it is a good idea to use specific number 64.5 in the transr_set_relation_embedder_dim method and 129 in torch.chunk(p_emb, 129, dim=1).
What we want is entity embedding (dim D), relation embedding (dim K) and projection matrix (dim K X D).
So, I think simply first change the dim of relation from K to K + K * D (= K * (1 + D)) in the transr_set_relation_embedder_dim method
and then divide p_emb into rel_emb and projection_matrix by using slicing like
```python
rel_emb = p_emb[:,K:]
projection_matrix = p_emb[:,:K]
```

What do you think about this?

Reply:
Yeah the hard-coded 64.5 & 129 is definitely not the way to go. I already changed it up a bit, see https://github.com/LJ171/kge/pull/2/commits/6688e828f3816000b5a4e4eb944a5ffda33a1420. It didn't occurred to me before that I can use 'config.get_default(rel_emb_conf_key + ".dim")' to retrieve the original k value and then apply the calculation as proposed by you. Thanks for the help 👍 implemented with https://github.com/LJ171/kge/pull/2/commits/500c0f7f01cba47a33f57a7ebb65ba08a1c6e792.